[SPARK-26432][CORE] Obtain HBase delegation token operation compatible with HBase 2.x.x version API

sujith71955 opened the pull request:
## What changes were proposed in this pull request?

While obtaining token from hbase service , spark uses deprecated API of hbase ,
```public static Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf)```
This deprecated API is already been removed from hbase 2.x version as part of the hbase 2.x major release. https://issues.apache.org/jira/browse/HBASE-14713_
there is one more stable API in 
```public static Token<AuthenticationTokenIdentifier> obtainToken(Connection conn)``` in TokenUtil class
spark shall use this stable api for getting the delegation token.

To invoke this api first connection object has to be retrieved from ConnectionFactory and the same connection can be passed to obtainToken(Connection conn) for getting token.
eg: Call   ```public static Connection createConnection(Configuration conf)```
, then call   ```public static Token<AuthenticationTokenIdentifier> obtainToken( Connection conn)```.

## How was this patch tested?
Manual testing is been done.
Manual test result:
Before fix:

![hbase-dep-obtaintok 1](https://user-images.githubusercontent.com/12999161/50699264-64cac200-106d-11e9-81b4-e50ae8097f27.png)

After fix:
1. Create 2 tables in hbase shell
 >Launch hbase shell
 >Enter commands to create tables and load data
    create 'table1','cf'
    put 'table1','row1','cf:cid','20'

    create 'table2','cf'
    put 'table2','row1','cf:cid','30'
	
 >Show values command
   get 'table1','row1','cf:cid'  will diplay value as 20
   get 'table2','row1','cf:cid'  will diplay value as 30
   
   
2.Run SparkHbasetoHbase class in testSpark.jar using spark-submit

spark-submit --master yarn-cluster --class com.mrs.example.spark.SparkHbasetoHbase --conf "spark.yarn.security.credentials.hbase.enabled"="true" --conf "spark.security.credentials.hbase.enabled"="true" --keytab /opt/client/user.keytab --principal sen testSpark.jar

The SparkHbasetoHbase test class will update the value of table2 with sum of values of table1 & table2.

table2 = table1+table2
As we can see in the snapshot the spark job has been successfully able to interact with hbase service and able to update the row count.
![obtaintok_success 1](https://user-images.githubusercontent.com/12999161/50699393-bd9a5a80-106d-11e9-96c6-6c250d561efa.png)

sujith71955 committed:
[SPARK-26432][CORE] Not able to connect Hbase 2.x service Getting NoSuchMethodException while trying to obtain token from Hbase 2.1 service.
## What changes were proposed in this pull request?

While obtaining token from hbase service , spark uses deprecated API of hbase public static Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf)
This deprecated API is already been removed from hbase 2.x version as part of the hbase 2.x major release.
there is one more stable API in public static Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) in TokenUtil class
spark shall use this stable api for getting the delegation token.
To invoke this api first connection object has to be retrieved from ConnectionFactory and the same connection can be passed to obtainToken(Connection conn) for getting token.
eg: Connection createConnection(Configuration conf) (), then call Token<AuthenticationTokenIdentifier> obtainToken(Connection conn)

## How was this patch tested?
To check the impact in current system existing test-cases and for verifying the feature manual testing is been done.
Test report has to be attached which i will update soon.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
ok to test

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -39,23 +41,46 @@ private[security] class HBaseDelegationTokenProvider
-      case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
+      // Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf) is a deprecated
+      // method and in Hbase 2.1.1 the method is already removed. there is one more public
```
Could you be more specific? Is this removed at 2.1.1?

sujith71955 replied:
shall we mention as 2.x, they removed it as part of Hbase 2.0.0 major version upgrade

sujith71955 replied:
https://issues.apache.org/jira/browse/HBASE-14713_ - JIRA for reference

sujith71955 replied:
updated the version. thnks .

SparkQA commented:
**[Test build #100649 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100649/testReport)** for PR 23429 at commit [`50b4733`](https://github.com/apache/spark/commit/50b47333f21cbf68ff4ef8812a1953851ab9feb8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6561/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6562/
Test PASSed.

SparkQA commented:
**[Test build #100651 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100651/testReport)** for PR 23429 at commit [`8c4b366`](https://github.com/apache/spark/commit/8c4b3662d22dfb274ec8a7f86a07724aa12dff9c).

vanzin commented:
Does your code keep support for HBase 1.x? Looks like it does not; that's the main reason why I just didn't submit my patch for this:
https://github.com/cloudera/spark/commit/3de80e469211797a099798e8bbe2b1fba74ff3df#diff-6e15cee083da1fefc4b9fa8a1f8642e9

It should be easy to keep both by just keeping the existing code in a different method, can calling that one if the 2.x method fails.

Also, please use the PR title to explain the solution, not the problem. Problems are described in the bug, not the PR.

SparkQA commented:
**[Test build #100651 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100651/testReport)** for PR 23429 at commit [`8c4b366`](https://github.com/apache/spark/commit/8c4b3662d22dfb274ec8a7f86a07724aa12dff9c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100651/
Test FAILed.

SparkQA commented:
**[Test build #100649 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100649/testReport)** for PR 23429 at commit [`50b4733`](https://github.com/apache/spark/commit/50b47333f21cbf68ff4ef8812a1953851ab9feb8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100649/
Test PASSed.

sujith71955 commented:
> HBase 1.x

The API which currently i am using is introduced in

sujith71955 committed:
[SPARK-26432][CORE] Not able to connect Hbase 2.x service Getting NoSuchMethodException while trying to obtain token from Hbase 2.1 service.
## What changes were proposed in this pull request?

While obtaining token from hbase service , spark uses deprecated API of hbase public static Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf)
This deprecated API is already been removed from hbase 2.x version as part of the hbase 2.x major release.
there is one more stable API in public static Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) in TokenUtil class
spark shall use this stable api for getting the delegation token.
To invoke this api first connection object has to be retrieved from ConnectionFactory and the same connection can be passed to obtainToken(Connection conn) for getting token.
eg: Connection createConnection(Configuration conf) (), then call Token<AuthenticationTokenIdentifier> obtainToken(Connection conn)

## How was this patch tested?
To check the impact in current system existing test-cases and for verifying the feature manual testing is been done.
Test report has to be attached which i will update soon.

sujith71955 commented:
> Does your code keep support for HBase 1.x? Looks like it does not; that's the main reason why I just didn't submit my patch for this:
> [cloudera/spark@3de80e4#diff-6e15cee083da1fefc4b9fa8a1f8642e9](https://github.com/cloudera/spark/commit/3de80e469211797a099798e8bbe2b1fba74ff3df#diff-6e15cee083da1fefc4b9fa8a1f8642e9)
> 
> It should be easy to keep both by just keeping the existing code in a different method, then calling that one if the 2.x method fails.
> 
> Also, please use the PR title to explain the solution, not the problem. Problems are described in the bug, not the PR.

Will update based on your comment. Thanks for the valuable feedback.

vanzin commented:
> The API which currently i am using obtainToken(Connection conn) is introduced in Hbase 0.98.9

Good to know. When I find some time I'll give this a try, just in case... it's hard to trust in reflection code without testing. :-/

sujith71955 commented:
> > The API which currently i am using obtainToken(Connection conn) is introduced in Hbase 0.98.9
> 
> Good to know. When I find some time I'll give this a try, just in case... it's hard to trust in reflection code without testing. :-/

Definitely you can give a try, anyways i will test and update the report in the PR, i already tested it with my current code and its working fine. Need to retest with  modified code. Thnks

vanzin commented:
Could you also remove the WIP from the title? Unless you don't actually want this to be pushed in the current form.

SparkQA commented:
**[Test build #100739 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100739/testReport)** for PR 23429 at commit [`d05448a`](https://github.com/apache/spark/commit/d05448af491b64010332043ba40fcb086ee68c3e).

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,70 @@ private[security] class HBaseDelegationTokenProvider
     } catch {
       case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
+        logWarning(s"Failed to get token from service $serviceName" + "Trying to invoke" +
```
`$serviceName" + "Trying` -> `$serviceName. Trying`?

sujith71955 replied:
handled, let me know if any more corrections required. thnks

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,70 @@ private[security] class HBaseDelegationTokenProvider
 
+  /**
+   * The HBase client API used in below method is introduced from HBase 0.98.9,1.0.0 version
+    * to invokethis api first connection object has to be retrieved from ConnectionFactory and the
```
indentation.

dongjoon-hyun replied:
`invokethis` -> `invoke this`

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,70 @@ private[security] class HBaseDelegationTokenProvider
+    * @param hadoopConf
+    * @param creds
+    */
+  def obtainDelegationTokensWithHBaseConn(hadoopConf: Configuration, creds: Credentials): Unit = {
```
`private def`?

sujith71955 replied:
done

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6634/
Test PASSed.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,70 @@ private[security] class HBaseDelegationTokenProvider
+    }
+  }
+
+  def getClassParamType(arrayOfMethod: Array[Method]): Class[_] = {
```
`private def`?

sujith71955 replied:
yes, should be

sujith71955 replied:
done. updated. thnks

SparkQA commented:
**[Test build #100741 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100741/testReport)** for PR 23429 at commit [`48d1bd2`](https://github.com/apache/spark/commit/48d1bd2d607ff068dd7f1ddec2c38b7a2a80fff1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6637/
Test PASSed.

SparkQA commented:
**[Test build #100744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100744/testReport)** for PR 23429 at commit [`0b05465`](https://github.com/apache/spark/commit/0b054650af6352f02fa2247efe94b7baab7d7564).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6640/
Test PASSed.

SparkQA commented:
**[Test build #100739 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100739/testReport)** for PR 23429 at commit [`d05448a`](https://github.com/apache/spark/commit/d05448af491b64010332043ba40fcb086ee68c3e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100739/
Test PASSed.

SparkQA commented:
**[Test build #100741 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100741/testReport)** for PR 23429 at commit [`48d1bd2`](https://github.com/apache/spark/commit/48d1bd2d607ff068dd7f1ddec2c38b7a2a80fff1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100741/
Test PASSed.

SparkQA commented:
**[Test build #100744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100744/testReport)** for PR 23429 at commit [`0b05465`](https://github.com/apache/spark/commit/0b054650af6352f02fa2247efe94b7baab7d7564).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100744/
Test PASSed.

srowen commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
+  }
+
+  private def getClassParamType(arrayOfMethod: Array[Method]): Class[_] = {
+    val seqOfFiltMethod = arrayOfMethod.filter(f => f.getName.equalsIgnoreCase("obtainToken"))
```
Nits: just write `filter(_.getName....)`. You can't ignore case here, right?

sujith71955 replied:
This Method is not required, as you told from reflection we can directly load the Class of the Connection  as below, we can omit the above logic.
  val connectionParamTypeClassRef = mirror.classLoader
        .loadClass("org.apache.hadoop.hbase.client.Connection")

srowen commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
+      if (null == obtainParamTypeClass) {
+        logDebug(s"Failed to invoke service to get token from service $serviceName")
+      }
+      else {
```
Nit: pull these up onto the previous line. Same for catch below.

sujith71955 replied:
done

srowen commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
+    val obtainHBaseConnection = mirror.classLoader.
+      loadClass("org.apache.hadoop.hbase.client.ConnectionFactory").
+      getMethod("createConnection", classOf[Configuration])
+    val conn = obtainHBaseConnection.invoke(null, hbaseConf(hadoopConf))
```
Can you cast this to at least `Closeable`? you do later. Is it known to be something more specific or does the type vary with version?

sujith71955 replied:
done

sujith71955 replied:
Yes, we can cast, no type shall not vary with version .

srowen commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
+    // method and in Hbase 2.0.0 the method is already removed. there is one more public
+    // consistent API Token<AuthenticationTokenIdentifier> obtainToken(Connection conn),
+    // in TokenUtil class which can be invoked using reflection.
+    val obtainHBaseConnection = mirror.classLoader.
```
What if these throw exceptions? later reflective access is wrapped in try-catch; is this different?

sujith71955 replied:
i had pulled down the statement inside try catch block, declared the connection variable above the block for later access.

srowen commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +53,71 @@ private[security] class HBaseDelegationTokenProvider
+    try {
+      val arrayOfMethod: Array[Method] = ClassLoader.getSystemClassLoader
+        .loadClass("org.apache.hadoop.hbase.security.token.TokenUtil").getMethods
+      val obtainParamTypeClass = getClassParamType(arrayOfMethod)
```
Could we rename the vars and methods here? I'm not clear what they do as they're named pretty generically. Are you looking for a method with a specific signature? the reflection API has methods to do that directly.

sujith71955 replied:
Logic not required now, remaining variable i renamed to more meaningful name, any suggestions welcome. thanks

sujith71955 committed:
[SPARK-26432][CORE] Not able to connect Hbase 2.x service Getting NoSuchMethodException while trying to obtain token from Hbase 2.1 service.
## What changes were proposed in this pull request?

While obtaining token from hbase service , spark uses deprecated API of hbase public static Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf)
This deprecated API is already been removed from hbase 2.x version as part of the hbase 2.x major release.
there is one more stable API in public static Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) in TokenUtil class
spark shall use this stable api for getting the delegation token.
To invoke this api first connection object has to be retrieved from ConnectionFactory and the same connection can be passed to obtainToken(Connection conn) for getting token.
eg: Connection createConnection(Configuration conf) (), then call Token<AuthenticationTokenIdentifier> obtainToken(Connection conn)

## How was this patch tested?
To check the impact in current system existing test-cases and for verifying the feature manual testing is been done.
Test report has to be attached which i will update soon.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6780/
Test PASSed.

SparkQA commented:
**[Test build #100887 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100887/testReport)** for PR 23429 at commit [`4c4a815`](https://github.com/apache/spark/commit/4c4a8156c6e3062025dd6b667e5dc34e44a13cf0).

sujith71955 commented:
@srowen @dongjoon-hyun @vanzin  Handled the comments and tested . thanks for the valuable suggestions. let me know for any clarifications or suggestions

SparkQA commented:
**[Test build #100887 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100887/testReport)** for PR 23429 at commit [`4c4a815`](https://github.com/apache/spark/commit/4c4a8156c6e3062025dd6b667e5dc34e44a13cf0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100887/
Test PASSed.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+   * same connection can be passed to
+   * Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) API
+   *
+   * @param hadoopConf
```
I would either add real documentation or drop the params.

sujith71955 replied:
updated the comment. thnks.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
+    val mirror = universe.runtimeMirror(Utils.getContextOrSparkClassLoader)
```
Why not having this under try umbrella?

sujith71955 replied:
right, we can move the statement inside. done.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+        // Seems to be spark is trying to get the token from HBase 2.x.x  version or above where the
+        // obtainToken(Configuration conf) API is been removed. Lets try obtaining the token from
+        // another compatible API of HBase service.
+        obtainDelegationTokensWithHBaseConn(hadoopConf, creds)
```
Wondering whether it would be better to try with the stable API first. Later it would be easier to removed 1.x support.

sujith71955 replied:
I think both way shall not make much difference, still we can take a call by getting opinion from other reviewers. thnks.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
       creds.addToken(token.getService, token)
     } catch {
       case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
```
Why dropped exception printing? It could still hold useful information if 2.x also failed.
Additionally wondering whether it's really a warning?

sujith71955 replied:
yes better to convey the error to  user, i modified the message . thnks

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+      creds.addToken(token.getService, token)
+    } catch {
+      case NonFatal(e) =>
+        logError(s"Failed to get token from service $serviceName", e)
```
The original level was warning, why error here?

sujith71955 replied:
I thought both the attempts of retrieving delegation token got failed, so basically spark will not able interact with hbase service in this scenario which will in-turn fail the user  attempt to do some operations based on hbase. so i thought better to log this as error instead of just warning. in  the scenario warning user can still successfully do his operations with some limitations at max. anyways i updated the level to error as the old code also was logging this failure as warning. Thanks.

SparkQA commented:
**[Test build #101123 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101123/testReport)** for PR 23429 at commit [`d0cb793`](https://github.com/apache/spark/commit/d0cb79377977dedb7b7c2c606a2b68a7a5f1236c).

SparkQA commented:
**[Test build #101123 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101123/testReport)** for PR 23429 at commit [`d0cb793`](https://github.com/apache/spark/commit/d0cb79377977dedb7b7c2c606a2b68a7a5f1236c).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101123/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6985/
Test PASSed.

sujith71955 commented:
@gaborgsomogyi thanks for the valuable feedback, let me know for any clarification you have still.
@vanzin @dongjoon-hyun @srowen  Please let me know if the above modifications are fine, or still any improvements required for the PR. thanks all for your valuable time.

SparkQA commented:
**[Test build #101124 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101124/testReport)** for PR 23429 at commit [`3e2091a`](https://github.com/apache/spark/commit/3e2091af653cbeb078523de0e8554a05de7f064c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6986/
Test PASSed.

SparkQA commented:
**[Test build #101126 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101126/testReport)** for PR 23429 at commit [`4881de3`](https://github.com/apache/spark/commit/4881de3665bbc7fdb4fa95d2741ac4497ea85dc2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6988/
Test PASSed.

SparkQA commented:
**[Test build #101124 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101124/testReport)** for PR 23429 at commit [`3e2091a`](https://github.com/apache/spark/commit/3e2091af653cbeb078523de0e8554a05de7f064c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #101126 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101126/testReport)** for PR 23429 at commit [`4881de3`](https://github.com/apache/spark/commit/4881de3665bbc7fdb4fa95d2741ac4497ea85dc2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101126/
Test PASSed.

sujith71955 commented:
@all 
Please let me know is any issues still needs to be addressed in this PR,  Without this fix  spark cannot be connected to hbase 2.x version which seems to be a major problem for hbase related usecases.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+        logWarning(s"Failed to get token from service $serviceName due to  " + e +
+          s" Retrying to fetch HBase security token with hbase connection parameter.")
+        // Seems to be spark is trying to get the token from HBase 2.x.x  version or above where the
+        // obtainToken(Configuration conf) API is been removed. Lets try obtaining the token from
```
Nit: is -> has

sujith71955 replied:
fixed. thanks

gaborgsomogyi reviewed (commented):
Other than the small thing looks good.
Maybe it's better not to use force push all the time because it's hard to follow the PR.

sujith71955 commented:
> Other than the small thing looks good.
> Maybe it's better not to use force push all the time because it's hard to follow the PR.

Valid point, will take care in future. thanks for suggestion

sujith71955 committed:
[SPARK-26432][CORE] Not able to connect Hbase 2.x service Getting NoSuchMethodException while trying to obtain token from Hbase 2.1 service.
## What changes were proposed in this pull request?

While obtaining token from hbase service , spark uses deprecated API of hbase public static Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf)
This deprecated API is already been removed from hbase 2.x version as part of the hbase 2.x major release.
there is one more stable API in public static Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) in TokenUtil class
spark shall use this stable api for getting the delegation token.
To invoke this api first connection object has to be retrieved from ConnectionFactory and the same connection can be passed to obtainToken(Connection conn) for getting token.
eg: Connection createConnection(Configuration conf) (), then call Token<AuthenticationTokenIdentifier> obtainToken(Connection conn)

## How was this patch tested?
To check the impact in current system existing test-cases and for verifying the feature manual testing is been done.
Test report has to be attached which i will update soon.

SparkQA commented:
**[Test build #101472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101472/testReport)** for PR 23429 at commit [`c1fed1e`](https://github.com/apache/spark/commit/c1fed1e5f73c9cac700dc985e53b546975f28fb2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7268/
Test PASSed.

SparkQA commented:
**[Test build #101472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101472/testReport)** for PR 23429 at commit [`c1fed1e`](https://github.com/apache/spark/commit/c1fed1e5f73c9cac700dc985e53b546975f28fb2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101472/
Test PASSed.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
   }
 
+  /**
+   * The HBase client API used in below method is introduced from HBase 0.98.9,1.0.0 version
```
0.98.9 or 1.0.0? Pick one.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+   * @param creds Credentials to add tokens and security keys to.
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
```
One argument per line.

sujith71955 replied:
updated. thanks

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+      // method and in Hbase 2.0.0 the method is already removed. there is one more public
+      // consistent API Token<AuthenticationTokenIdentifier> obtainToken(Connection conn),
+      // in TokenUtil class which can be invoked using reflection.
+      val connectionFactoryClass = mirror.classLoader.
```
Move all `.` to the next line.

sujith71955 replied:
done. thanks

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+    var connectionHbaseClient = null.asInstanceOf[Closeable]
+    try {
+      val mirror = universe.runtimeMirror(Utils.getContextOrSparkClassLoader)
+      // Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf) is a deprecated
```
Already explained in the method comment.

sujith71955 replied:
updated. thanks

vanzin commented on core/src/main/scala/org/apache/spark/deploy/security/HBaseDelegationTokenProvider.scala:
```diff
@@ -50,12 +52,59 @@ private[security] class HBaseDelegationTokenProvider
+   */
+  private def obtainDelegationTokensWithHBaseConn(
+      hadoopConf: Configuration, creds: Credentials): Unit = {
+    var connectionHbaseClient = null.asInstanceOf[Closeable]
```
`var connectionHbaseClient: Closeable = null`

Also `hbaseConnection` or just `connection` is a better name.

sujith71955 replied:
updated. thanks

vanzin reviewed (commented):
I tested this on old hbase and it works; but I thought you were going to keep only the new code, after you said that it was compatible with old hbase.

sujith71955 committed:
[SPARK-26432][CORE] Not able to connect Hbase 2.x service Getting NoSuchMethodException while trying to obtain token from Hbase 2.1 service.
## What changes were proposed in this pull request?

While obtaining token from hbase service , spark uses deprecated API of hbase public static Token<AuthenticationTokenIdentifier> obtainToken(Configuration conf)
This deprecated API is already been removed from hbase 2.x version as part of the hbase 2.x major release.
there is one more stable API in public static Token<AuthenticationTokenIdentifier> obtainToken(Connection conn) in TokenUtil class
spark shall use this stable api for getting the delegation token.
To invoke this api first connection object has to be retrieved from ConnectionFactory and the same connection can be passed to obtainToken(Connection conn) for getting token.
eg: Connection createConnection(Configuration conf) (), then call Token<AuthenticationTokenIdentifier> obtainToken(Connection conn)

## How was this patch tested?
To check the impact in current system existing test-cases and for verifying the feature manual testing is been done.
Test report has to be attached which i will update soon.

sujith71955 commented:
> I tested this on old hbase and it works; but I thought you were going to keep only the new code, after you said that it was compatible with old hbase.

Thanks a lot for testing vanzin, even i tested and attached the report above please refer.
I updated based on the below remarks, and i think it make sense as this particular API is been  introduced only  in 0.98.9 version. for older versions this approach will help us.Hope its fine, let me know for any suggestions.
![image](https://user-images.githubusercontent.com/12999161/51611013-33defe00-1f44-11e9-97a6-1bdfe933e1ca.png)

SparkQA commented:
**[Test build #101590 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101590/testReport)** for PR 23429 at commit [`2e38f71`](https://github.com/apache/spark/commit/2e38f711084f3900e14e240629502c829f6e7f3d).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7358/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7359/
Test PASSed.

sujith71955 commented:
> Test FAILed.
> Refer to this link for build results (access rights to CI server needed):
> https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7358/
> Test FAILed.

Seems to be unknown error, can someone re-trigger the build. thanks

SparkQA commented:
**[Test build #101590 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101590/testReport)** for PR 23429 at commit [`2e38f71`](https://github.com/apache/spark/commit/2e38f711084f3900e14e240629502c829f6e7f3d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101590/
Test FAILed.

gaborgsomogyi commented:
retest this please

SparkQA commented:
**[Test build #101599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101599/testReport)** for PR 23429 at commit [`2e38f71`](https://github.com/apache/spark/commit/2e38f711084f3900e14e240629502c829f6e7f3d).

vanzin commented:
looks good pending tests (which really shouldn't be affected anyway...)

SparkQA commented:
**[Test build #101599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101599/testReport)** for PR 23429 at commit [`2e38f71`](https://github.com/apache/spark/commit/2e38f711084f3900e14e240629502c829f6e7f3d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101599/
Test PASSed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #101697 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101697/testReport)** for PR 23429 at commit [`2e38f71`](https://github.com/apache/spark/commit/2e38f711084f3900e14e240629502c829f6e7f3d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7438/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101697/
Test PASSed.

vanzin commented:
Merging to master.
